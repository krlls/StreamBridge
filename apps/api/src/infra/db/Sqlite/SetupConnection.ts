import { DataSource, EntityTarget, Repository } from 'typeorm'
import Database from 'better-sqlite3'

import { NODE_ENV, sqlLite } from '../../../config'

export const dataSource = new DataSource(sqlLite[NODE_ENV])

export class SqliteDB {
  private static _instance: SqliteDB
  private dbConnect!: DataSource
  private testdb!: any

  public static get instance(): SqliteDB {
    if (!this._instance) this._instance = new SqliteDB()

    return this._instance
  }

  async setupTestDB() {
    this.testdb = new Database('')
    this.dbConnect = await dataSource.initialize()
  }

  async teardownTestDB() {
    await this.dbConnect.destroy()
    this.testdb.close()
  }
}

export const getRepository = <Entity extends Record<string, any>>(target: EntityTarget<Entity>): Repository<Entity> => {
  return dataSource.getRepository(target)
}

export default dataSource
