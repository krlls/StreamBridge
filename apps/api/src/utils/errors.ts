import { ServiceError, ServiceResultDTO } from '../types/common'
import { Api } from '../types/TApi'

export const isServiceError = <T extends object>(result: ServiceResultDTO<T>): result is ServiceError =>
  'error' in result

export const isSpotifyAuthError = (obj: any): obj is Api.Streaming.Token.ErrorReq => Object.hasOwn(obj, 'error')
