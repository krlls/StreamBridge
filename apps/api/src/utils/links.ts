import { serverConfig } from '../config'

export const createPatch = (...args: string[]) => args.join('')

export const apiLink = (...args: string[]) => createPatch(serverConfig.apiUrl, ...args)
