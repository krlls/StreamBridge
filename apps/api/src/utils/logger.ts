/*eslint-disable no-console*/
import moment from 'moment'
import clc from 'cli-color'
import { RouterContext } from 'koa-router'
import { isEmpty } from 'lodash'

import { serverConfig } from '../config'
import { EStreamingType } from '../types/common'

enum ELogType {
  WARN = 'WARN',
  ERROR = 'ERROR',
  INFO = 'INFO',
}

enum ELogger {
  API = 'API',
  STREAMING = 'STREAMING CLIENT',
}

abstract class Logger {
  protected static logInfo(name: string, ...args: any) {
    return this.log(ELogType.INFO, clc.green(name), ...args)
  }

  protected static logWarn(name: string, ...args: any) {
    return this.log(ELogType.WARN, clc.green(name), ...args)
  }

  protected static logError(name: string, ...args: any) {
    return this.log(ELogType.ERROR, clc.green(name), ...args)
  }

  private static log(type: ELogType, name: string, ...args: any) {
    if (serverConfig.silent) {
      return
    }

    const color = {
      [ELogType.WARN]: clc.yellow,
      [ELogType.ERROR]: clc.red,
      [ELogType.INFO]: clc.blue,
    }[type]

    return console.log(`[${moment().format('YY-MM-DD hh:mm:ss:SS')}] ${color(type)} [${name}]`, ...args)
  }
}

export class ApiLogger extends Logger {
  static info(ctx: RouterContext, ...args: any) {
    return this.logInfo(ELogger.API, ctx.req.method, ctx.request.url, ...args)
  }

  static warn(ctx: RouterContext, ...args: any) {
    return this.logWarn(ELogger.API, ctx.req.method, ctx.request.url, ...args)
  }

  static error(ctx: RouterContext, ...args: any) {
    return this.logError(ELogger.API, ctx.req.method, ctx.request.url, ...args)
  }
}

export class StreamingLogger extends Logger {
  constructor(private name: EStreamingType) {
    super()
  }
  info(...args: any) {
    return StreamingLogger.logInfo(this.name, ...args)
  }

  warn(...args: any) {
    return StreamingLogger.logWarn(this.name, ...args)
  }

  error(...args: any) {
    return StreamingLogger.logError(this.name, ...args)
  }
}

export class ImportLogger extends Logger {
  static info(...args: any) {
    return StreamingLogger.logInfo('IMPORT', ...args)
  }

  static warn(...args: any) {
    return StreamingLogger.logWarn('IMPORT', ...args)
  }

  static error(...args: any) {
    return StreamingLogger.logError('IMPORT', ...args)
  }
}

export const showNotEmpty = <A, B>(a: A, b: B) => (!isEmpty(a) ? a : !isEmpty(b) ? b : '')
