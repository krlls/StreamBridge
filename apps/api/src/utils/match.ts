import { JaroWinklerDistance } from 'natural'

// https://en.wikipedia.org/wiki/Jaro%E2%80%93Winkler_distance
export const calcMatchScore = (s1: string, s2: string) => JaroWinklerDistance(s1, s2, { ignoreCase: true })
