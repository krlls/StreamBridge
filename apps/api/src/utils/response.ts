import { RouterContext } from 'koa-router'

export const respond200 = (ctx: RouterContext) => {
  ctx.status = 200
  ctx.body = ''
}

export const respond204 = <A extends string, B>(ctx: RouterContext, body?: Record<A, B>) => {
  ctx.status = 204
  ctx.body = body || ''
}

export const respond200json = <T extends Record<any, any>>(ctx: RouterContext, body?: T) => {
  ctx.type = 'json'
  ctx.body = body
}

export const respond200plain = (ctx: RouterContext, body: string) => {
  ctx.status = 200
  ctx.type = 'text/plain'
  ctx.body = body
}

export const respond200Html = (ctx: RouterContext, body: string) => {
  ctx.status = 200
  ctx.type = 'text/html'
  ctx.body = body
}

export const respond403 = <A extends string, B>(ctx: RouterContext, body?: Record<A, B> | string) => {
  ctx.status = 403
  ctx.body = body || ''
}

export const respond400 = <A extends string, B>(ctx: RouterContext, body?: Record<A, B> | string) => {
  ctx.status = 400
  ctx.body = body || ''
}

export const respond409 = <A extends string, B>(ctx: RouterContext, body?: Record<A, B> | string) => {
  ctx.status = 409
  ctx.body = body || ''
}

export const respond401json = <A extends string, B>(ctx: RouterContext, body?: Record<A, B>) => {
  ctx.status = 401
  ctx.body = body
}

export const respond404 = <A extends string, B>(ctx: RouterContext, body?: Record<A, B> | string) => {
  ctx.status = 404
  ctx.body = body || ''
}

export const respond500 = (ctx: RouterContext) => {
  ctx.status = 500
  ctx.body = ''
}
