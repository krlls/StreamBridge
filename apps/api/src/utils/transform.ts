import { Api } from 'api-types'

import { EStreamingType } from '../types/common'

export const convertStreamingName = (name: string): EStreamingType | undefined =>
  ({
    [Api.Streaming.EApiStreamingType.SPOTIFY]: EStreamingType.SPOTIFY,
  })[name]
