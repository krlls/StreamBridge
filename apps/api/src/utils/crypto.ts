import * as argon2 from 'argon2'
import jwt from 'koa-jwt'
import * as jsonwebtoken from 'jsonwebtoken'
import { inject, injectable } from 'inversify'
import { RouterContext } from 'koa-router'
import Koa from 'koa'

import { serverConfig } from '../config'
import { TYPES } from '../types/const'
import { IUserService } from '../modules/user/interfaces/IUserService'
import { isServiceError } from './errors'
import { respond401json } from './response'

export const ALGORITHM = 'HS256'

export const hashPass = async (pass: string) => argon2.hash(pass)
export const verifyPass = async (pass: string, hash: string) => argon2.verify(hash, pass)

const checkAuth = jwt({
  secret: serverConfig.jwtSecret,
  algorithms: [ALGORITHM],
})

export const checkStreamingToken = async (token: string) =>
  new Promise<boolean>((resolve) =>
    jsonwebtoken.verify(token, serverConfig.jwtStreamingAuthSecret, { algorithms: [ALGORITHM] }, (err) => {
      if (err) {
        return resolve(false)
      }

      return resolve(true)
    }),
  )

export const getDataFromToken = async <T extends Record<string, string | number>>(token: string) =>
  new Promise<T | undefined>((resolve) =>
    jsonwebtoken.verify(token, serverConfig.jwtStreamingAuthSecret, { algorithms: [ALGORITHM] }, (err, resp) => {
      if (err) {
        return resolve(undefined)
      }

      return resolve(resp as T)
    }),
  )

export const createSignedJwt = <T extends object>(payload: T) =>
  new Promise<string | null>((resolve) =>
    jsonwebtoken.sign(
      payload,
      serverConfig.jwtStreamingAuthSecret,
      {
        expiresIn: serverConfig.jwtStreamingAuthExpires,
        algorithm: ALGORITHM,
      },
      (error, encoded) => {
        if (error || !encoded) {
          return resolve(null)
        }

        return resolve(encoded)
      },
    ),
  )

@injectable()
export class AuthChecker {
  @inject(TYPES.UserService) private userService: IUserService

  createMiddleware() {
    return [checkAuth, this.checkUser]
  }

  private checkUser: Koa.Middleware = async (ctx, next) => {
    const user = await this.userService.findUserById(ctx.state.user?.userId)

    if (isServiceError(user)) {
      return respond401json(ctx as RouterContext, user)
    }

    return next()
  }
}
