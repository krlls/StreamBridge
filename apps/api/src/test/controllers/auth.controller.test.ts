import { expect, it, describe, beforeEach, afterEach } from '@jest/globals'
import { Api } from 'api-types'

import { TestApp } from '../index.test'
import { SqliteDB } from '../../infra/db/Sqlite/SetupConnection'
import { authUrl, testUserData, userUrl } from '../helpers/test.helpers'
import { createSignedJwt } from '../../utils/crypto'
import { testsWithToken } from '../helpers/testMany'

describe('Auth controllers tests', () => {
  beforeEach(async () => {
    await SqliteDB.instance.setupTestDB()
    await TestApp.post(userUrl(Api.User.Create.URL)).send(testUserData)
  })

  afterEach(async () => {
    await SqliteDB.instance.teardownTestDB()
  })

  it('Auth works', async () => {
    const authData: Api.Auth.Login.Req = {
      login: testUserData.login,
      pass: testUserData.pass,
    }
    const response = await TestApp.post(authUrl(Api.Auth.Login.URL)).send(authData)

    expect(response.status).toBe(200)
    expect(JSON.parse(response.text)).toHaveProperty('token')
  })

  it('Auth with not valid password', async () => {
    const authData: Api.Auth.Login.Req = {
      login: testUserData.login,
      pass: testUserData.pass + 'random string',
    }
    const response = await TestApp.post(authUrl(Api.Auth.Login.URL)).send(authData)

    expect(response.status).toBe(403)
    expect(JSON.parse(response.text)).toHaveProperty('error')
  })

  it('Auth with not valid login', async () => {
    const authData: Api.Auth.Login.Req = {
      login: testUserData.login + 'random string',
      pass: testUserData.pass,
    }
    const response = await TestApp.post(authUrl(Api.Auth.Login.URL)).send(authData)

    expect(response.status).toBe(403)
    expect(JSON.parse(response.text)).toHaveProperty('error')
    expect(response.text).toMatchSnapshot()
  })

  it('Auth with not valid data', async () => {
    const authData = { login: '12', password: testUserData.pass }
    const response = await TestApp.post(authUrl(Api.Auth.Login.URL)).send(authData)

    expect(response.status).toBe(400)
  })

  it('Test endpoints with wrong user id', async () => {
    const userToken = await createSignedJwt({ userId: 89 })

    return testsWithToken(userToken || '')()
  })

  it('Test endpoints with zero user id', async () => {
    const userToken = await createSignedJwt({ userId: 0 })

    return testsWithToken(userToken || '')()
  })
  it('Test endpoints with empty user id', async () => {
    const userToken = await createSignedJwt({})

    return testsWithToken(userToken || '')()
  })

  it('Test endpoints with undefined user id', async () => {
    const userToken = await createSignedJwt({ userId: undefined })

    return testsWithToken(userToken || '')()
  })
})
