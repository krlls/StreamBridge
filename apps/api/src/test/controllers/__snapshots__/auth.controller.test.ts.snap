// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth controllers tests Auth with not valid login 1`] = `"{"error":"Inkorrect credentials"}"`;
